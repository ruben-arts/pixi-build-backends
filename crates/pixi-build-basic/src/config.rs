use std::path::{Path, PathBuf};

use indexmap::IndexMap;
use pixi_build_backend::generated_recipe::BackendConfig;
use serde::{Deserialize, Serialize};

#[derive(Debug, Default, Deserialize, Serialize, Clone)]
#[serde(rename_all = "kebab-case", deny_unknown_fields)]
pub struct BasicBackendConfig {
    /// The bash or cmd commands that should be run in the build.
    #[serde(default)]
    pub script: Vec<String>,
    /// Extra args for CMake invocation
    #[serde(default)]
    pub extra_args: Vec<String>,
    /// Environment Variables
    #[serde(default)]
    pub env: IndexMap<String, String>,
    /// If set, internal state will be logged as files in that directory
    pub debug_dir: Option<PathBuf>,
    /// Extra input globs to include in addition to the default ones
    #[serde(default)]
    pub extra_input_globs: Vec<String>,
}

impl BackendConfig for BasicBackendConfig {
    fn debug_dir(&self) -> Option<&Path> {
        self.debug_dir.as_deref()
    }

    /// Merge this configuration with a target-specific configuration.
    /// Target-specific values override base values using the following rules:
    /// - extra_args: Platform-specific completely replaces base
    /// - env: Platform env vars override base, others merge
    /// - debug_dir: Not allowed to have target specific value
    /// - extra_input_globs: Platform-specific completely replaces base
    fn merge_with_target_config(&self, target_config: &Self) -> miette::Result<Self> {
        if target_config.debug_dir.is_some() {
            miette::bail!("`debug_dir` cannot have a target specific value");
        }

        Ok(Self {
            script: if target_config.script.is_empty() {
                self.script.clone()
            } else {
                target_config.script.clone()
            },
            extra_args: if target_config.extra_args.is_empty() {
                self.extra_args.clone()
            } else {
                target_config.extra_args.clone()
            },
            env: {
                let mut merged_env = self.env.clone();
                merged_env.extend(target_config.env.clone());
                merged_env
            },
            debug_dir: self.debug_dir.clone(),
            extra_input_globs: if target_config.extra_input_globs.is_empty() {
                self.extra_input_globs.clone()
            } else {
                target_config.extra_input_globs.clone()
            },
        })
    }
}
