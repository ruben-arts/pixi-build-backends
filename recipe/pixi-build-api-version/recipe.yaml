# yaml-language-server: $schema=https://raw.githubusercontent.com/prefix-dev/recipe-format/main/schema.json
context:
  name: pixi-build-api-version
  version: "1"

package:
  name: ${{ name }}
  version: ${{ version }}

build:
  noarch: generic

about:
  homepage: https://github.com/prefix-dev/pixi-build-backends
  summary: The protocol version for pixi build backends.
  description: |
    This package is used to define the API version of the protocol used by pixi build backends.
  license: BSD-3-Clause
  documentation: https://prefix-dev.github.io/pixi-build-backends
  repository: https://github.com/prefix-dev/pixi-build-backends
