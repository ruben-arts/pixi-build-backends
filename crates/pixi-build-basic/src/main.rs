mod config;

use std::{collections::BTreeSet, path::Path, sync::Arc};

use config::BasicBackendConfig;
use miette::IntoDiagnostic;
use pixi_build_backend::{
    generated_recipe::{GenerateRecipe, GeneratedRecipe, PythonParams},
    intermediate_backend::IntermediateBackendInstantiator,
};
use recipe_stage0::recipe::Script;

#[derive(Default, Clone)]
pub struct BasicGenerator {}

impl GenerateRecipe for BasicGenerator {
    type Config = BasicBackendConfig;

    fn generate_recipe(
        &self,
        model: &pixi_build_types::ProjectModelV1,
        config: &Self::Config,
        manifest_root: std::path::PathBuf,
        host_platform: rattler_conda_types::Platform,
        _python_params: Option<PythonParams>,
    ) -> miette::Result<GeneratedRecipe> {
        let mut generated_recipe = GeneratedRecipe::from_model(model.clone());

        generated_recipe.recipe.build.script = Script {
            content: config.script.clone(),
            env: config.env.clone(),
            ..Default::default()
        };

        Ok(generated_recipe)
    }

    fn extract_input_globs_from_build(
        config: &Self::Config,
        _workdir: impl AsRef<Path>,
        _editable: bool,
    ) -> BTreeSet<String> {
        [].iter()
            .map(|s: &&str| s.to_string())
            .chain(config.extra_input_globs.clone())
            .collect()
    }
}

#[tokio::main]
pub async fn main() {
    if let Err(err) = pixi_build_backend::cli::main(|log| {
        IntermediateBackendInstantiator::<BasicGenerator>::new(log, Arc::default())
    })
    .await
    {
        eprintln!("{err:?}");
        std::process::exit(1);
    }
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use indexmap::IndexMap;
    use pixi_build_backend::protocol::ProtocolInstantiator;
    use pixi_build_types::{
        ProjectModelV1,
        procedures::{conda_outputs::CondaOutputsParams, initialize::InitializeParams},
    };
    use rattler_build::console_utils::LoggingOutputHandler;

    use super::*;

    #[test]
    fn test_input_globs_includes_extra_globs() {
        let config = BasicBackendConfig {
            extra_input_globs: vec!["custom/*.c".to_string()],
            ..Default::default()
        };

        let result = BasicGenerator::extract_input_globs_from_build(&config, PathBuf::new(), false);

        insta::assert_debug_snapshot!(result);
    }
}
