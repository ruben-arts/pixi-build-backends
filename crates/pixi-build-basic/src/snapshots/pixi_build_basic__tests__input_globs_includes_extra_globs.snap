---
source: crates/pixi-build-basic/src/main.rs
expression: result
snapshot_kind: text
---
{
    "custom/*.c",
}
